core/src/avm1/object/script_object.rs

```diff
@@ -589,78 +550,30 @@ impl<'gc> TObject<'gc> for ScriptObject<'gc> {
         self.0.as_ptr() as *const ObjectPtr
     }
 
-    fn length(&self, _activation: &mut Activation<'_, 'gc, '_>) -> Result<i32, Error<'gc>> {
-        match &self.0.read().array {
-            ArrayStorage::Vector(vector) => Ok(vector.len() as i32),
-            ArrayStorage::Properties { length } => Ok(*length),
-        }
+    fn length(&self, activation: &mut Activation<'_, 'gc, '_>) -> Result<i32, Error<'gc>> {
+        self.get_data("length", activation)
+            .coerce_to_i32(activation)
     }
 
     fn set_length(
         &self,
         activation: &mut Activation<'_, 'gc, '_>,
         new_length: i32,
     ) -> Result<(), Error<'gc>> {
-        let to_remove = match &mut self.0.write(activation.context.gc_context).array {
-            ArrayStorage::Vector(vector) => {
-                if new_length >= 0 {
-                    let old_length = vector.len();
-                    let new_length = new_length as usize;
-                    vector.resize(new_length, Value::Undefined);
-                    Some(new_length..old_length)
-                } else {
-                    None
-                }
-            }
-            ArrayStorage::Properties { length } => {
-                *length = new_length;
-                None
-            }
-        };
-        if let Some(to_remove) = to_remove {
-            for i in to_remove {
-                self.sync_native_property(
-                    &i.to_string(),
-                    activation.context.gc_context,
-                    None,
-                    true,
-                );
+        if let Value::Number(old_length) = self.get_data("length", activation) {
+            for i in new_length.max(0)..f64_to_wrapping_i32(old_length) {
+                self.delete_element(activation, i);
             }
         }
-        self.sync_native_property(
-            "length",
-            activation.context.gc_context,
-            Some(new_length.into()),
-            false,
-        );
-        Ok(())
+        self.set_data("length", new_length.into(), activation)
```

Comment:
This is more of a "future PR" change, but I think that perhaps `set_length` should only exist as a function inside `ArrayObject`. It's a little odd to me that this is inside the standard `ScriptObject` (only an Array should care about erasing elements based on length).

The only other place it's used right now is `AsBroadcaster`, but that also should be manually calling array methods instead:
```
var o = {};
AsBroadcaster.initialize(o);
// Swap out the AsBroadcaster's listener with our own fake array.
o._listeners = {length: 0, push: function() { trace("Called push"); }};
o.addListener(listener); // "Called push!" in Flash, nothing in Ruffle
```

Reply:
Good point. For now I corrected `AsBroadcaster` to call `push` directly.

Reply:
Differing the `set_length` overload on `ScriptObject` vs `ArrayObject` works, but I still wonder if it's better to keep this logic entirely `ArrayObject` or `globals::array`, because only `Array` methods should be using this call -- setting or not setting length can be handled with an `as_array_object` check. This also encourages us to implement and go through a proper interface when interacting with arrays such as the `AsBroadcaster` case, as opposed messing with the array internals directly.